File: file_system_tool/backend/src/api/schemas/filesystem.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from enum import Enum


class FileType(str, Enum):
    FILE = "file"
    DIRECTORY = "directory"


class CreateFileRequest(BaseModel):
    path: str = Field(..., description="Full path including filename")
    size: int = Field(..., ge=0, description="File size in bytes")


class CreateDirectoryRequest(BaseModel):
    path: str = Field(..., description="Full path for directory")


class DeleteItemRequest(BaseModel):
    path: str = Field(..., description="Path to delete")
    recursive: bool = Field(default=False, description="Recursive delete for directories")


class FileInfo(BaseModel):
    name: str
    path: str
    type: FileType
    size: Optional[int] = None
    created: datetime
    modified: datetime
    inode_number: Optional[int] = None
    blocks: Optional[List[int]] = None


class DirectoryListing(BaseModel):
    path: str
    items: List[FileInfo]


class DiskInfo(BaseModel):
    total_blocks: int
    block_size: int
    total_capacity_mb: float
    used_blocks: int
    free_blocks: int
    usage_percentage: float


class BlockStatus(BaseModel):
    block_number: int
    is_allocated: bool
    owner_inode: Optional[int] = None
    owner_file: Optional[str] = None


class DiskVisualization(BaseModel):
    total_blocks: int
    blocks_per_row: int
    block_states: List[Dict[str, Any]]  # {block_num, status, color, owner}
    statistics: DiskInfo

File: file_system_tool/frontend/src/types/filesystem.ts
export type FileType = 'file' | 'directory';

export interface FileInfo {
  name: string;
  path: string;
  type: FileType;
  size?: number;
  created: string;
  modified: string;
  inode_number?: number;
  blocks?: number[];
}

export interface DirectoryListing {
  path: string;
  items: FileInfo[];
}

export interface DiskInfo {
  total_blocks: number;
  block_size: number;
  total_capacity_mb: number;
  used_blocks: number;
  free_blocks: number;
  usage_percentage: number;
}

export interface BlockState {
  block_number: number;
  status: 'free' | 'allocated';
  color: string;
  owner_inode?: number;
  owner_file?: string;
}

export interface DiskVisualization {
  total_blocks: number;
  blocks_per_row: number;
  block_states: BlockState[];
  statistics: DiskInfo;
}

export interface CrashReport {
  crash_id: number;
  crash_type: string;
  timestamp: string;
  affected_blocks: number[];
  severity: string;
  recoverable: boolean;
  description: string;
}

export interface RecoveryReport {
  success: boolean;
  recovered_transactions: number;
  rolled_back_transactions: number;
  recovery_time: number;
  errors: string[];
}

export interface DefragmentationReport {
  success: boolean;
  files_defragmented: number;
  fragmentation_before: number;
  fragmentation_after: number;
  blocks_moved: number;
  time_taken: number;
}

export interface PerformanceMetrics {
  timestamp: string;
  disk_usage_percentage: number;
  fragmentation_percentage: number;
  cache_hit_rate: number;
  read_throughput_mbps: number;
  write_throughput_mbps: number;
  iops: number;
  free_space_mb: number;
}

export interface LogEntry {
  timestamp: string;
  level: 'debug' | 'info' | 'warning' | 'error' | 'success';
  message: string;
}

export interface WebSocketMessage {
  type: string;
  data?: any;
  message?: string;
}
